postgis/lwgeom_hexgrid.c

```diff
@@ -0,0 +1,360 @@
+Datum ST_Square(PG_FUNCTION_ARGS);
+Datum ST_ShapeGrid(PG_FUNCTION_ARGS);
+
+/* ********* ********* ********* ********* ********* ********* ********* ********* */
+
+/* XXX TO DO: what about negative sizes? */
+
+/* Ratio of hexagon edge length to height = 2*cos(pi/6) */
+#define HXR 1.7320508075688774
+
+static const double hex_x[] = {0.0, 1.0, 1.5,     1.0, 0.0, -0.5,     0.0};
+static const double hex_y[] = {0.0, 0.0, 0.5*HXR, HXR, HXR,  0.5*HXR, 0.0};
+
+static LWGEOM *
+hexagon(double origin_x, double origin_y, double edge, int cell_i, int cell_j, uint32_t srid)
+{
+	double height = edge * HXR;
+	POINT4D pt;
+	POINTARRAY **ppa = lwalloc(sizeof(POINTARRAY*));
+	POINTARRAY *pa = ptarray_construct(0, 0, 7);
+	uint32_t i;
+
+	double offset_x = origin_x + (1.5 * edge * cell_i);
+	double offset_y = origin_y + (height * cell_j) + (0.5 * height * (cell_i % 2));
+	pt.z = pt.m = 0.0;
+
+	for (i = 0; i < 7; ++i)
+	{
+		pt.x = edge * hex_x[i] + offset_x;
+		pt.y = edge * hex_y[i] + offset_y;
+		ptarray_set_point4d(pa, i, &pt);
+	}
+
+	ppa[0] = pa;
+	return lwpoly_as_lwgeom(lwpoly_construct(srid, NULL, 1 /* nrings */, ppa));
+}
+
+static int
+hexagon_grid_size(double edge, const GBOX *gbox, int *cell_width, int *cell_height)
+{
+	double hex_height = edge * HXR;
+	double bounds_width  = gbox->xmax - gbox->xmin;
+	double bounds_height = gbox->ymax - gbox->ymin;
+	int width  = (int)floor(bounds_width/(1.5*edge)) + 1;
+	int height = (int)floor(bounds_height/hex_height) + 1;
+	if (cell_width) *cell_width = width;
+	if (cell_height) *cell_height = height;
+	return width * height;
+}
+
+/* ********* ********* ********* ********* ********* ********* ********* ********* */
+
+static const double sqr_x[] = {0.0, 1.0, 1.0, 0.0, 0.0};
+static const double sqr_y[] = {0.0, 0.0, 1.0, 1.0, 0.0};
+
+static LWGEOM *
+square(double origin_x, double origin_y, double E, int cell_i, int cell_j, uint32_t srid)
+{
+	POINT4D pt;
+	POINTARRAY **ppa = lwalloc(sizeof(POINTARRAY*));
+	POINTARRAY *pa = ptarray_construct(0, 0, 5);
+	uint32_t i;
+
+	double offset_x = origin_x + (E * cell_i);
+	double offset_y = origin_y + (E * cell_j);
+	pt.z = pt.m = 0.0;
+
+	for (i = 0; i < 5; ++i)
+	{
+		pt.x = E * sqr_x[i] + offset_x;
+		pt.y = E * sqr_y[i] + offset_y;
+		ptarray_set_point4d(pa, i, &pt);
+	}
+
+	ppa[0] = pa;
+	return lwpoly_as_lwgeom(lwpoly_construct(srid, NULL, 1 /* nrings */, ppa));
+}
+
+static int
+square_grid_size(double edge, const GBOX *gbox, int *cell_width, int *cell_height)
+{
+	int width  = (int)floor((gbox->xmax - gbox->xmin)/edge) + 1;
+	int height = (int)floor((gbox->ymax - gbox->ymin)/edge) + 1;
+	if (cell_width) *cell_width = width;
+	if (cell_height) *cell_height = height;
+	return width * height;
+}
+
+/* ********* ********* ********* ********* ********* ********* ********* ********* */
+
+typedef enum
+{
+	SHAPE_SQUARE,
+	SHAPE_HEXAGON,
+	SHAPE_TRIANGLE
+} GeometryShape;
+
+typedef struct GeometryGridState
+{
+	int cell_current;
```

Comment:
It's personal preference, but I'd rather use int32_t / int64_t to know for sure the size of the int.

Reply:
And I've just noticed that since cell_current can be int32_t x int32_t size, it might overflow if it's just 32 bytes.

Reply:
done